### PR TITLE
core utils implementation using go

### PR DESCRIPTION
**Pull request by eyad-hussein**
The following coreutils are implemented:
- head
 - tail
 - wc (word count)
 - cat
 - echo
 - tree
- env
 - true
 - false

Acceptance Criteria
- Each utility is implemented as a separate Go program.
- Includes basic error handling.
- Includes a README with build and usage instructions for all utilities.

### CODE REVIEW

**Comment by Eslam-Nawara** [comment]
the name of the pr should reflect it's content, `Project is fully done, made sure to follow the acceptance criteria after the session's feedback` not a good name

**Comment by eyad-hussein** [comment]
Would this work?
```Coreutils implemented based on acceptance criteria```

**Comment by Eslam-Nawara** [comment]
@eyad-hussein `based on acceptance criteria` can be added to the description if you want, `core utils implementation` should be enough

**Comment by eyad-hussein** [comment]
@Eslam-Nawara understood,  thank you

File: README.md [hunk 1]
```diff
@@ -1 +1,252 @@
...
+            ├── tree.go
+            ├── true.go
+            ├── wc.go
+            └── yes.go
+```
+- **cmd/**: Holds commands directories and files.
+- **pkg/**: Holds package files.
+- **src/Coreutils-EYAD_HUSSEIN/**: Source directory for the project.
+
+## Commands Implemented
```
Comment (maayarosama) [inline README.md:53]: please provide a description to every flag

File: src/Coreutils-EYAD_HUSSEIN/pkg/cat.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+	var nFlag bool
+
+	flag.BoolVar(&nFlag, "n", false, "number all output lines")
+
+	flag.Parse()
+
+	ok := len(flag.Args()) > 0
+
+	if !ok {
```
Comment (maayarosama) [inline src/Coreutils-EYAD_HUSSEIN/pkg/cat.go]:
```suggestion
	if !(len(flag.Args()) > 0) {
```

File: src/Coreutils-EYAD_HUSSEIN/pkg/cat.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+func Cat() {
+
+	var nFlag bool
+
+	flag.BoolVar(&nFlag, "n", false, "number all output lines")
+
+	flag.Parse()
+
+	ok := len(flag.Args()) > 0
```
Comment (maayarosama) [inline src/Coreutils-EYAD_HUSSEIN/pkg/cat.go]: the variable can be removed and use the condition directly in the if condition

File: README.md (hunk 1 above)
Comment (eyad-hussein) [reply README.md:53]: Done

File: src/Coreutils-EYAD_HUSSEIN/pkg/cat.go (hunk 3 above)
Comment (eyad-hussein) [reply src/Coreutils-EYAD_HUSSEIN/pkg/cat.go]: done, replaced each occurance with the condition

File: src/Coreutils-EYAD_HUSSEIN/pkg/tree.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func Tree() {
+
+	var maxDepth int
+
+	flag.IntVar(&maxDepth, "L", -1, "Descend only level directories deep")
```
Comment (maayarosama) [inline src/Coreutils-EYAD_HUSSEIN/pkg/tree.go:15]: why -1?

File: src/Coreutils-EYAD_HUSSEIN/pkg/env.go [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+	"fmt"
+	"os"
+	"strings"
+)
+
+func Env() {
+
+	for _, env := range os.Environ() {
+		envPair := strings.Split(env, "=")
+		fmt.Println(envPair[0], envPair[1])
```
Comment (maayarosama) [inline src/Coreutils-EYAD_HUSSEIN/pkg/env.go]: no need to split, just print env as it is

**Review by maayarosama — CHANGES_REQUESTED** [review]

File: src/Coreutils-EYAD_HUSSEIN/pkg/tree.go (hunk 4 above)
Comment (eyad-hussein) [reply src/Coreutils-EYAD_HUSSEIN/pkg/tree.go:15]: this way, given no flag, the function will never return from the first base case, which is

```
if maxDepth != -1 && depth > maxDepth {
		return
	}
```

This way, the recursion tree won't stop until every file and directory in each level inside whatever entry point is given is printed(expect for hidden ones of course(second base case)) --> all levels are traversed and printed, which is the behavior of tree when no flag is passed.

Any other positive number or zero would not have worked.

File: src/Coreutils-EYAD_HUSSEIN/pkg/env.go (hunk 5 above)
Comment (eyad-hussein) [reply src/Coreutils-EYAD_HUSSEIN/pkg/env.go]: Done

**Review by maayarosama — APPROVED** [review]

**Pull request merged**